seaworthy/tests-core/test_logs.py

```diff
@@ -255,12 +252,76 @@ def test_streaming_waits(self):
         t0 = datetime.now()
         self.assertEqual(self.stream(con), [b'hello\n', b'goodbye\n'])
         t1 = datetime.now()
-        self.assertEqual(self.stream(con), [b'hello\n', b'goodbye\n'])
+        self.assertEqual(self.stream(con), [])
         t2 = datetime.now()
         self.assertLess(0.3, (t1 - t0).total_seconds())
         self.assertLess((t2 - t1).total_seconds(), 0.3)
 
 
+class TestStreamWithHistoryFunc(unittest.TestCase):
+    def mkcontainer(self, *args, **kw):
+        con = FakeLogsContainer(*args, **kw)
+        self.addCleanup(con.cleanup)
+        return con
+
+    def stream(self, con, timeout=1):
+        # Always clean up the feeder machinery when we're done. This is only
+        # necessary if we timed out, but it doesn't hurt to clean up an
+        # already-clean FakeLogsContainer.
+        try:
+            return list(stream_logs(con, timeout=timeout))
+        finally:
+            con.cancel_feeder()
+
+    def swh(self, con, timeout=0.5, **kw):
+        # Always clean up the feeder machinery when we're done. This is only
+        # necessary if we timed out, but it doesn't hurt to clean up an
+        # already-clean FakeLogsContainer.
+        try:
+            return stream_with_history(con, timeout=timeout, **kw)
+        finally:
+            con.cleanup()
+
+    def test_stream_only(self):
+        """
+        If there are no historical logs, we get all the streamed logs.
+        """
+        con = self.mkcontainer([
+            (0.1, b'hello\n'),
+            (0.1, b'goodbye\n'),
+        ])
+        self.assertEqual(list(self.swh(con)), [b'hello\n', b'goodbye\n'])
+
+    def test_historical_only(self):
+        """
+        If all the logs have been streamed, we only get the old ones.
+        """
+        con = self.mkcontainer([
+            (0.1, b'hello\n'),
+            (0.1, b'goodbye\n'),
+        ])
+        self.assertEqual(self.stream(con), [b'hello\n', b'goodbye\n'])
+        self.assertEqual(self.stream(con), [])
+        self.assertEqual(list(self.swh(con)), [b'hello\n', b'goodbye\n'])
+
+    def test_timeout_and_stream_again(self):
+        """
+        If we take too long to get the next line, we time out. When we stream
+        again, we only get historical logs as well as new ones.
```

Comment:
Small nit: the word "only" here doesn't seem to make sense?

Reply:
Copy/paste fail. :-)